lib/core/src/Ouroboros/Network/Client/Wallet.hs

```diff
@@ -475,62 +494,97 @@ type LocalStateQueryResult state = Either AcquireFailure state
 --        │       ┌───────────┴───┐             ┌────────┴───────┐
 --        └───────┤   Acquired    │────────────>│   Querying     │
 --                └───────────────┘             └────────────────┘
+-- FIXME: I think the diagram is wrong. It is possible to query multiple times
+-- while we're acquired.
 --
+-- NOTE: Using AnyCardanoEra arguably goes against the grain of the abstract
+-- block type. We might be able to use the (Header block) as replacemet for
+-- (AnyCardanoEra, Tip block).
 localStateQuery
-    :: forall m block . (MonadThrow m, MonadSTM m)
-    => TQueue m (LocalStateQueryCmd block m)
+    :: forall m block . (MonadIO m, MonadThrow m, MonadSTM m, MonadTimer m, Eq (Point block))
+    => TVar m (AnyCardanoEra, Tip block)
+    -> TQueue m (LocalStateQueryCmd block m)
         -- ^ We use a 'TQueue' as a communication channel to drive queries from
         -- outside of the network client to the client itself.
         -- Requests are pushed to the queue which are then transformed into
         -- messages to keep the state-machine moving.
     -> LocalStateQueryClient block (Point block) (Query block) m Void
-localStateQuery queue =
+localStateQuery tipVar queue =
     LocalStateQueryClient clientStIdle
   where
     clientStIdle
         :: m (LSQ.ClientStIdle block (Point block) (Query block) m Void)
-    clientStIdle = awaitNextCmd <&> \case
-        CmdQueryLocalState pt query respond ->
-            LSQ.SendMsgAcquire pt (clientStAcquiring query respond)
+    clientStIdle = do
+        cmd <- awaitNextCmd
+        (era, pt) <- second getTipPoint <$> atomically (readTVar tipVar)
+        pure $ LSQ.SendMsgAcquire pt (clientStAcquiring cmd pt era 0)
 
     clientStAcquiring
-        :: forall state. Query block state
-        -> (LocalStateQueryResult state -> m ())
+        :: LocalStateQueryCmd block m
+        -> Point block
+        -> AnyCardanoEra
+        -> Int
         -> LSQ.ClientStAcquiring block (Point block) (Query block) m Void
-    clientStAcquiring query respond = LSQ.ClientStAcquiring
-        { recvMsgAcquired = clientStAcquired query respond
-        , recvMsgFailure = \failure -> do
-                respond (Left failure)
-                clientStIdle
+    clientStAcquiring qry oldPt era retries = LSQ.ClientStAcquiring
+        { recvMsgAcquired = clientStAcquired qry era
+        , recvMsgFailure = \_failure -> do
+            (newEra, newPt) <- second getTipPoint <$> (atomically $ do
+                t <- readTVar tipVar
+                guard (getTipPoint (snd t) /= oldPt)
+                return t)
+            pure $ LSQ.SendMsgAcquire newPt (clientStAcquiring qry newPt newEra (retries + 1))
```

Comment:
This is counting retries, what for? Is the goal to cap that at some point?

Reply:
Initially I capped it, but now it is not capped and this should be removed.